### PR TITLE
Developer msft

### PR DESCRIPTION
**Pull request by MeritedHobbit**
1. Added Compliance Level Tags. 
2. Added some exception handling. 

image

### CODE REVIEW

**Review by afsanac-nv — COMMENTED** [review]
Minor comments/questions added. 

I've tried running this version in my end and is seeing following error.
```
~/ocp-diag-ctam/ctam$ python3 ctam.py -w ../workspace/
WorkSpace : ../workspace/
Test case BasicTelemetryTestGroup not found.
Test Run Failed: {
    "ctam.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/ctam.py",
        "lineno": 208,
        "name": "main"
    },
    "test_runner.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/test_runner.py",
        "lineno": 345,
        "name": "run"
    },
    "type": "AttributeError",
    "message": "'NoneType' object has no attribute 'tags'"
}

Test exited with status code*: FAIL - Test failed due to exception: 'NoneType' object has no attribute 'tags'
Log Directory: None

*Note: Return/Status Codes - PASS(0): All tests passed, FAIL(1): Execution/runtime failure or test failure
```

My test_runner.json looks like this:
```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "Need to add one",
    "title": "Test Runner Configuration",
    "description": "Configure the TestRunner obj, list of test cases overrides a test suite",
    "output_override_directory": "",
    "debug_mode": true,
    "console_log": true,
    "progress_bar":  true,
    "include_tags": ["L0"],
    "exclude_tags": [],
    "test_sequence" : [],
    "group_sequence" : [],
    "active_test_suite": "dev_test_suite",
    "dev_test_suite": ["F0","F1", "T0"],
    "full_compliance_test_suite": [],
    "regression_test_suite_0": ["F0","F1", "F63", "F64", "F8", "F88", "T0"],
    "regression_test_suite_1": ["F0", "F1", "F23", "F26", "F28", "F27", "F22", "F25", "F18", "F32"],
    "regression_test_suite_2": ["F0", "F1", "F16", "F19", "F55", "F62", "F89"],
    "test_uri_response_excel":"Excel file name for checking the response"
}
```

The log about "...skipped due to..." is coming from line 355 in test_runner.py (`elif self.test_sequence` condition).

How to run using tags? Anything wrong in my test_runner.json?

**Comment by afsanac-nv** [comment]
I've also tried running `regression_test_suite_0` with following test_runner.json, but am still getting error.
```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "Need to add one",
    "title": "Test Runner Configuration",
    "description": "Configure the TestRunner obj, list of test cases overrides a test suite",
    "output_override_directory": "",
    "debug_mode": true,
    "console_log": true,
    "progress_bar":  true,
    "include_tags": [],
    "exclude_tags": [],
    "test_sequence" : [],
    "group_sequence" : [],
    "active_test_suite": "regression_test_suite_0",
    "dev_test_suite": [],
    "full_compliance_test_suite": [],
    "regression_test_suite_0": ["F0","F1", "F63", "F64", "F8", "F88", "T0"],
    "regression_test_suite_1": ["F0", "F1", "F23", "F26", "F28", "F27", "F22", "F25", "F18", "F32"],
    "regression_test_suite_2": ["F0", "F1", "F16", "F19", "F55", "F62", "F89"],
    "test_uri_response_excel":"Excel file name for checking the response"
}
```

```
~/ocp-diag-ctam/ctam$ python3 ctam.py -w ../workspace/
WorkSpace : ../workspace/
PING 
...
...
...
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.784/0.784/0.784/0.000 ms
{
    "TimeStamp": "11-21-2023T16:17:47",
    "Message": "Binded port 18888"
},
Test Run Failed: {
    "ctam.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/ctam.py",
        "lineno": 208,
        "name": "main"
    },
    "test_runner.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/test_runner.py",
        "lineno": 574,
        "name": "_run_group_test_cases"
    },
    "_fw_update_group_N.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/tests/fw_update/fw_update_group_N/_fw_update_group_N.py",
        "lineno": 57,
        "name": "teardown"
    },
    "test_group.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/tests/test_group.py",
        "lineno": 79,
        "name": "test_run"
    },
    "type": "NotImplementedError",
    "message": "need to call TestGroup.SetUpAssociations"
}

Test exited with status code*: FAIL - Test failed due to exception: need to call TestGroup.SetUpAssociations
Log Directory: None

*Note: Return/Status Codes - PASS(0): All tests passed, FAIL(1): Execution/runtime failure or test failure
```

**Comment by v-sourabhsingh** [comment]
> Collaborator

We are working on tag functionality but for now if you want to run tag then you have to pass test sequence, group sequence or active_test_suite to make it work.
Along with this, add the tag to the particular group of the test case.

File: ctam/ctam.py [hunk 1]
```diff
@@ -196,13 +196,15 @@ def main():
                 sequence_group_override=args.group_sequence,
             )
         else:
+            all_tests = test_hierarchy.get_all_tests()
```
Comment (afsanac-nv) [inline ctam/ctam.py:199]: This is overriding test_runner.json options. Please fix it.

**Pull request closed without merging**